great_ape_safe/ape_api/ebtc.py

```diff
@@ -866,3 +876,277 @@ def authority_set_authority(self, new_authority):
+
+    def _assert_debt_balance(self, debt_amount):
+        total_debt_bal = self.ebtc_token.balanceOf(self.safe.address)
+        assert total_debt_bal >= debt_amount
+        return total_debt_bal
+
+    def _assert_cdp_id_ownership(self, cdp_id):
+        cdp_safe_owned = self.sorted_cdps.getCdpsOf(self.safe.address)
+        assert cdp_id in cdp_safe_owned
+
+    def _assert_health_new_icr(self, new_icr):
+        # NOTE: chose 120% to have alert with a breath range from MCR
+        assert new_icr > self.SAFE_ICR_THRESHOLD
+
+    def _hint_helper_values(self, coll_amount, debt_amount):
+        nicr = self.hint_helpers.computeNominalCR(coll_amount, debt_amount)
+
+        total_cdps = self.sorted_cdps.getSize()
+        hint, _, _ = self.hint_helpers.getApproxHint(
+            nicr, total_cdps, 42
+        )  # random seed 42
+
+        upper_hint, lower_hint = self.sorted_cdps.findInsertPosition(nicr, hint, hint)
+
+        return upper_hint, lower_hint
+
+    def open_cdp(self, coll_amount, target_cr):
+        """
+        @dev Opens a new cdp position. Attention to target_cr unit format, see below!
+        @param coll_amount The total stETH collateral amount deposited for the specified Cdp.
+        @param target_cr The desired target collateral ratio in the cdp position. Unit format: cr follows a 10^18 formatting
+        """
+        # verify: is it coll balance available
+        self._assert_collateral_balance(coll_amount)
+
+        debt_balance_before = self.ebtc_token.balanceOf(self.safe.address)
+
+        # 1. collateral approval for BO
+        self.collateral.approve(self.borrower_operations.address, coll_amount)
+
+        # 2. decide borrow amount based on: collateral, feed price & CR
+        feed_price = self.price_feed.fetchPrice.call()
+        borrow_amount = (coll_amount * feed_price) / target_cr
+
+        # 3. open cdp with args
+        upper_hint, lower_hint = self._hint_helper_values(coll_amount, borrow_amount)
+        cdp_id = self.borrower_operations.openCdp(
+            borrow_amount, upper_hint, lower_hint, coll_amount
+        ).return_value
+
+        # 4. assertions:
+
+        # 4.1. verify cdp id is owned by caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        # 4.2. debt balance of caller increased (difference)
+        bal_diff = self.ebtc_token.balanceOf(self.safe.address) - debt_balance_before
+        assert bal_diff == borrow_amount
+
+        return cdp_id
+
+    def close_cdp(self, cdp_id):
+        """
+        @dev Closes a target cdp id.
+        @param cdp_id The CdpId on which this operation is operated.
+        """
+        # verify: cdp id ownership from caller
+        self._assert_cdp_id_ownership(cdp_id)
+
+        # verify: sufficient ebtc is hold for closing
+        (
+            cdp_id_debt,
+            cdp_id_coll,
+            _,
+            cdp_id_liquidator_reward_shares,
+            _,
+            _,
+        ) = self.cdp_manager.Cdps(cdp_id)
```

Comment:
here for `cdp_id_debt` and `cdp_id_coll`, better to use `cdp_manager.getSyncedDebtAndCollShares(cdp_id)`

https://github.com/ebtc-protocol/ebtc/blob/main/packages/contracts/contracts/CdpManagerStorage.sol#L745C14-L745C40

```
(cdp_id_debt, cdp_id_coll) = self.cdp_manager.getSyncedDebtAndCollShares(cdp_id)
```

Reply:
suggestion added in 4fcce46cf1cb8a85bbef7007a50b6af49de60e10